service/matching/matcher.go

```diff
@@ -162,19 +274,79 @@ func (tm *TaskMatcher) PollForQuery(ctx context.Context) (*internalTask, error)
 	case <-ctx.Done():
 		tm.scope().IncCounter(metrics.PollTimeoutCounter)
 		return nil, ErrNoTasks
+	case token := <-tm.fwdrPollReqTokenC():
+		if task, err := tm.fwdr.ForwardPoll(ctx); err == nil {
+			token.release()
+			return task, nil
+		}
+		token.release()
+		return tm.pollForQuery(ctx)
 	}
 }
 
 // UpdateRatelimit updates the task dispatch rate
 func (tm *TaskMatcher) UpdateRatelimit(rps *float64) {
-	tm.limiter.UpdateMaxDispatch(rps)
+	if rps == nil {
+		return
+	}
+	rate := *rps
+	nPartitions := tm.numPartitions()
+	if rate > float64(nPartitions) {
+		// divide the rate equally across all partitions
+		rate = rate / float64(tm.numPartitions())
```

Comment:
probably `numpartitions-1` if it is greater than 1.. as the parent partition should not do any work in a fully loaded system when the limit will actually apply..

Reply:
> as the parent partition should not do any work in a fully loaded system when the limit will actually apply
This isn't true unless I am misunderstanding your statement. Parent partition is just another partition and when the system is fully loaded, it gets roughly 1/N of the traffic

Reply:
Because we try local sync match first on a fully loaded system I thought tasks would rarely make it to the parent. but i see the parent is treated as an equivalent partition in the `pick` logic (i guess parent could starve otherwise), so yes `1/N` should be fine